### ISSUE TITLE
Extend EQDSK interface to output B0

### ISSUE THREAD

**Issue by krystophny** [itpplasma/libneo#36]
### Result

Python routines that evaluate

1. Magnetic field in `R,Z`
2. `R,Z` for given `s, theta_geom`
3. Combination to get `B(s,theta_geom)`

Think about how to separate or combine EQDSK RZ and flux coordinate (Boozer) modules/classes.

**Comment by marjohma** [comment]
@krystophny issue already exists.

**Comment by krystophny** [comment]
*Automated triage pass (Claude), grounded in the repository at `fe3f76e`. Verify before acting.*

Still relevant; two of the three requested items are missing.

The issue asks for three things. What exists today is a **prerequisite** for item 2, not item 2 itself: the *forward* map `(R,Z) -> (s_pol, theta_geom)` landed in `882aeb6` "feat(eqdsk): add R,Z to flux coordinate interpolation and Sphinx docs (#254)" (2026-01-15). Status per item:

| Issue item | State |
|---|---|
| 1. `B(R,Z)` | **missing** — no B-field evaluator anywhere in `python/libneo/` |
| 2. `(R,Z)` for given `(s, theta_geom)` (*inverse* map) | **missing** — only the forward map exists |
| 3. `B(s, theta_geom)` | **missing** — follows from 1 + 2 |
| "separate or combine RZ and Boozer modules" | **unanswered** — proposed answer under step 6 |

The one comment ("issue already exists") does not resolve to any tracked issue asking for this. The nearest neighbours are libneo#33 "Add checks between Fortran and Python EQDSK routines", libneo#360 "efit_to_boozer: tested, clean version with clear COCOS convention", and libneo#362 "Extract separatrix value from EQDSK file" — none of them asks for a Python B-field evaluator. Not a duplicate. Spec below.

## Purpose

Every downstream analysis task that maps a diagnostic signal or an orbit onto an equilibrium needs `|B|` and its components at an arbitrary point, and needs to go *back* from a flux label to real space to place a probe, a ray, or a plotting grid. libneo has this in Fortran (`field_eq`, `src/magfie/field_divB0.f90:127`) but the Python side stops at the poloidal flux, and `field_eq` is not reachable from Python: `src/f2py_interfaces/f2py_magfie.f90` exposes only `init_vmec`.

The result is that the same Grad-Shafranov algebra gets re-derived ad hoc wherever it is needed — see `python/libneo/eqdsk_to_boozer_chartmap.py:271-272`, which open-codes `B_R = -dpsi_dZ / R_amp; B_Z = dpsi_dR / R_amp` inline inside the chartmap converter, and `python/scripts/draft_magfie_gfile.py:36-46`, a loose unshipped draft that does roughly what this issue asks via a dead `cffi` shim (`libmc_fffi`, not present in the repo). Closing this gap gives a pure-Python, dependency-light path to `B` from a g-file, usable without building the Fortran library.

## Current state

`python/libneo/eqdsk.py` defines `class eqdsk_file` with, in order:

- `_build_psi_interpolator` (`:8`) — builds a `scipy.interpolate.RegularGridInterpolator` over `(Z, R)` with `method='cubic'`, and caches `_psi_at_axis`/`_psi_at_edge` with a COCOS sign guard (`:14-19`).
- `psi_at_rz` (`:21`), `spol_at_rz` (`:59`), `theta_geometric_at_rz` (`:82`), `rz_to_flux_coords` (`:109`).

That is the whole coordinate API; it is also exactly the member list exported in `docs/eqdsk.rst` (`:members: psi_at_rz, spol_at_rz, theta_geometric_at_rz, rz_to_flux_coords`) and covered by `test/python/test_eqdsk_interpolation.py`.

The blocking detail is the interpolator choice:

```python
  def _build_psi_interpolator(self):
    from scipy.interpolate import RegularGridInterpolator
    import numpy as np
    self._psi_interpolator = RegularGridInterpolator(
        (self.Z, self.R), self.PsiVs, method='cubic', bounds_error=False
    )
```

`RegularGridInterpolator` exposes no derivatives, and `B_pol` is a derivative of `psi`. This object must be replaced by (or supplemented with) a `RectBivariateSpline`, which supports `dx=`/`dy=` — the pattern already used at `python/libneo/eqdsk_to_boozer_chartmap.py:249-253`.

The needed data is already parsed in `generate_from_file`:

- `self.fprof` = `F(psi) = R*B_tor` on `self.psieqprof` (uniform 0..1 normalized flux), `:212-215`
- `self.Btor_at_R0` (`:191`), `self.R0` (`:156`), `self.Rpsi0`/`self.Zpsi0` (`:181-183`)
- `self.PsiVs`, reshaped to `(nzgr, nrgr)` at `:242`
- `self.qprof` (`:245-248`)
- `self.Lcfs`, read at `:260-262`, reshaped at `:264` and transposed at `:265` to final shape `(2, npbound)` — so `Lcfs[0]` is R and `Lcfs[1]` is Z

The reference implementation to match is `field_eq`, `src/magfie/field_divB0.f90:244-265`:

```fortran
  Brad = -dpsidz/rrr        ! :244
  Bzet =  dpsidr/rrr        ! :245
  ...
  if(use_fpol) then         ! :257
    psihat=psif/psi_sep
    if(psihat.gt.1.d0) then ! :259
      fpol=splfpol(0,nrad)  ! :260
      fpol_prime=0.d0
    else
      call splint_fpol(psihat,fpol,fpol_prime)
    end if
    Bphi = fpol/rrr         ! :265
```

Note that `field_eq` works in CGS — units are converted at `src/magfie/field_divB0.f90:195-200` (`rad*1.d2` cm, `psi*1.d8`, `btf*1.d4` G) — and normalizes with `psihat = psif/psi_sep`, which is *not* the same as the Python `s_pol = (psi - psi_axis)/(psi_edge - psi_axis)`. The Python layer stays SI throughout, so cross-checks need `B[T] = B[G]*1e-4`, and only the *vacuum freeze* behaviour of `field_eq` is being mirrored, not its normalization.

On the Boozer side, `python/efit_to_boozer/boozer.py:42` `get_angles_G_and_B0(s, num_theta)` returns, per surface, `r, th_boozers, th_geoms, theta_symflux, Gs, B0mod` (`:97`) — i.e. `|B|` *and* `theta_geom` parameterized by the symmetry-flux angle — but only via the full Fortran field-line-integration pipeline and with `s = s_tor`, not `s_pol`. That is the "separate or combine" question in the issue body, answered under step 6.

## Reproduction / evidence
```bash
cd /path/to/libneo
pip install -e ".[dev]"          # the [dev] extra exists, pyproject.toml:26
python - <<'EOF'
from libneo import eqdsk_file
eq = eqdsk_file("test/resources/input_efit_file.dat")
print(eq.rz_to_flux_coords(eq.Rpsi0 + 0.2, eq.Zpsi0))
print(hasattr(eq, "b_at_rz"), hasattr(eq, "rz_at_flux_coords"))
EOF
```

Observed on `fe3f76e`:

```
(np.float64(0.26841577640463404), 0.0)
False False
```

Expected after this issue: `b_at_rz` and `rz_at_flux_coords` present, with `BR, Bphi, BZ = eq.b_at_rz(eq.Rpsi0, eq.Zpsi0)` giving `Bphi` close to `eq.Btor_at_R0 * eq.R0 / eq.Rpsi0` and a vanishing poloidal field.

The tolerances below were checked against a throwaway prototype on this fixture, so they are known to be reachable, with margin:

| Quantity on `test/resources/input_efit_file.dat` | Measured | Criterion |
|---|---|---|
| `Bphi` on axis vs `Btor_at_R0*R0/Rpsi0` (both ≈ `-2.405`) | rel. err `2.1e-4` | `< 2e-2` |
| `hypot(BR,BZ)/abs(Bphi)` on axis | `6.5e-7` | `< 1e-2` |
| inverse-map round trip, 4 surfaces × 17 angles | 0/68 failures | see below |

**Fixture:** `test/resources/input_efit_file.dat` — the same g-file `test/python/test_eqdsk_interpolation.py:11` already uses. Note that `python/tests/test.geqdsk` is *not* usable: it is a dangling symlink to `../../../../data/MASTU/EQDSK/MAST_47051_210ms.geqdsk`, a machine-local path outside the checkout. Do not write tests against it.

**Independent cross-check against Fortran:** run the `efit_to_boozer` f2py module the way `python/libneo/eqdsk_to_boozer_chartmap.py:191` does (write `field_divB0.inp` pointing at the g-file, then call `magdata`), and compare `B0mod` from `get_angles_G_and_B0` against `hypot(BR, BZ, Bphi)` from the new `b_at_flux_coords`. This is the concrete content of libneo#33.

## Behaviour (BDD)

```gherkin
Scenario: Toroidal field on axis matches the g-file header
  Given the g-file test/resources/input_efit_file.dat loaded as eqdsk_file
  When b_at_rz(eq.Rpsi0, eq.Zpsi0) is called
  Then Bphi is within 2% of eq.Btor_at_R0 * eq.R0 / eq.Rpsi0
  And hypot(BR, BZ) is below 1e-2 * abs(Bphi)

Scenario: Vacuum toroidal field outside the LCFS falls off as 1/R
  Given the same equilibrium
  When b_at_rz is evaluated at (R, eq.Zpsi0) for R in linspace(2.18, eq.R[-2], 20)
  Then Bphi * R equals eq.fprof[-1] to within 1e-9 relative at every point
  # 2.18 > max(eq.Lcfs[0]) = 2.1308, so every sample is outside the LCFS
  # (verified: min s_pol over the 20 samples is 1.1889);
  # eq.R[-2] = 2.6400 stays inside the psi grid box, so no NaN.

Scenario: Inverse flux map round-trips
  Given s_pol in {0.1, 0.3, 0.6, 0.9} and theta_geom in linspace(-pi, pi, 17, endpoint=False)
  When R, Z = eq.rz_at_flux_coords(s_pol, theta_geom) then s2, t2 = eq.rz_to_flux_coords(R, Z)
  Then abs(s2 - s_pol) < 1e-6
  And the wrapped difference abs(t2 - theta_geom) < 1e-10
  And no call raises

Scenario: B in flux coordinates is wired to B in cylindrical coordinates
  Given s_pol = 0.5 and theta_geom = 0.7
  When BR, Bphi, BZ = eq.b_at_flux_coords(0.5, 0.7)
  And R, Z = eq.rz_at_flux_coords(0.5, 0.7)
  Then b_at_rz(R, Z) equals (BR, Bphi, BZ) exactly
  # b_at_flux_coords is defined as b_at_rz(*rz_at_flux_coords(...)), so this is
  # a composition/wiring guard, not an independent numerical check.

Scenario: Python B agrees with Fortran field_eq
  Given the efit_to_boozer f2py module initialised on the same g-file
  When |B| is evaluated on 32 points of the s_tor = 0.5 surface
  Then max relative deviation from B0mod * 1e-4 (G -> T) is below 1e-3
```

## Implementation plan

1. **`python/libneo/eqdsk.py`** — rewrite `_build_psi_interpolator` (`:8-19`) to build
   `self._psi_spline = RectBivariateSpline(self.R, self.Z, self.PsiVs.T, kx=3, ky=3)`
   (note the transpose: `PsiVs` is stored `(nzgr, nrgr)`, see `:242`; with this argument order `dx=1` is `d/dR` and `dy=1` is `d/dZ`).
   Keep `psi_at_rz` (`:21`) behaviour and signature byte-identical, including the `grid=` semantics and the scalar-unwrap, and keep the `_psi_at_axis`/`_psi_at_edge` sign guard exactly as at `:14-19`, so `test/python/test_eqdsk_interpolation.py` stays green unchanged.

2. **`python/libneo/eqdsk.py`** — add `dpsi_at_rz(self, R, Z)` returning `(dpsi_dR, dpsi_dZ)` via `self._psi_spline(R, Z, dx=1, grid=False)` and `dy=1`. Units: Wb/rad per metre, matching the file.

3. **`python/libneo/eqdsk.py`** — add `_build_fpol_interpolator(self)` caching `CubicSpline(np.asarray(self.psieqprof), self.fprof)` (data at `:212-215`; `psieqprof` is a plain Python list), plus `fpol_at_rz(self, R, Z)`: evaluate `s_pol = spol_at_rz(R, Z)` and return the spline at `np.clip(s_pol, 0.0, 1.0)`. Clamp on **both** sides: interpolation noise can push `s_pol` slightly negative near the axis, and `psieqprof` starts at exactly 0. The upper clamp reproduces the vacuum freeze of `src/magfie/field_divB0.f90:257-261` (`fpol=splfpol(0,nrad)` once `psihat > 1`), giving the constant `self.fprof[-1]` outside the LCFS.

4. **`python/libneo/eqdsk.py`** — add `b_at_rz(self, R, Z)` returning `(B_R, B_phi, B_Z)` in Tesla:
   `B_R = -dpsi_dZ / R`, `B_Z = dpsi_dR / R`, `B_phi = fpol_at_rz(R, Z) / R`.
   Vectorize over array input and unwrap scalars the same way `psi_at_rz` does. Add `bmod_at_rz(self, R, Z) = sqrt(B_R**2 + B_phi**2 + B_Z**2)`. Document in the docstring that the sign convention is that of the file as read (GEQDSK `sgn_cyl = +1`, as asserted at `src/magfie/geqdsk_tools.f90:269`), and that COCOS standardisation is the job of the `geqdsk_standardise` subroutine at `src/magfie/geqdsk_tools.f90:307` (driven by the `standardise_geqdsk` program in `src/magfie/standardise_geqdsk.f90`) — do **not** silently re-sign here. On this fixture `Bphi` on axis is negative (`-2.405 T`); compare signed, do not take absolute values.

5. **`python/libneo/eqdsk.py`** — add `rz_at_flux_coords(self, s_pol, theta_geom)`. Implementation: precompute, from `self.Lcfs` (`:260-265`), the polar description of the LCFS about the magnetic axis —
   `thl = arctan2(Lcfs[1] - Zpsi0, Lcfs[0] - Rpsi0)`, `ll = hypot(Lcfs[0] - Rpsi0, Lcfs[1] - Zpsi0)`, sorted by `thl`. Then for each `theta_geom`, set `l_max = 0.98 * np.interp(theta_geom, thl, ll, period=2*np.pi)`, define
   `f(l) = spol_at_rz(Rpsi0 + l*cos(theta), Zpsi0 + l*sin(theta)) - s_pol`,
   and solve with `scipy.optimize.brentq(f, 1e-6, l_max, xtol=1e-12)`. Return `(R, Z)`. Raise `ValueError` naming the offending `(s_pol, theta_geom)` if `s_pol <= 0` or `f` does not change sign across the bracket. This bracket was verified to succeed for all 68 `(s_pol, theta_geom)` combinations of the round-trip scenario on the test fixture.
   Add `b_at_flux_coords(self, s_pol, theta_geom)` = `b_at_rz(*rz_at_flux_coords(s_pol, theta_geom))`.

6. **Design answer to "separate or combine"** — keep them separate, and record this in `docs/eqdsk.rst`: `eqdsk_file` is the *cylindrical + geometric-flux* layer (pure Python, no Fortran, label `s_pol`, angle `theta_geom`); `python/efit_to_boozer/boozer.py` stays the *Boozer* layer (Fortran-backed, label `s_tor`, angle `theta_boozer`). The bridge between the two labels is `FluxConverter` in `python/libneo/flux_converter.py` (`spol2stor` at `:59`, `stor2spol` at `:66`). Do not add an `s_tor` accessor to `eqdsk_file`; add instead a documented one-liner example using
   `FluxConverter(eq.qprof, axis_polflux=eq.PsiaxisVs, edge_polflux=eq.PsiedgeVs)` —
   `FluxConverter` defaults `spol_profile` to `linspace(0, 1, len(q_profile))`, which matches the grid `eq.qprof` is stored on.

7. **`python/libneo/eqdsk_to_boozer_chartmap.py`** — deduplicate the inline field algebra: the `psi_spl = RectBivariateSpline(...)` block at `:249-253` and the open-coded `B_R = -dpsi_dZ / R_amp; B_Z = dpsi_dR / R_amp` at `:271-272`.
   **Read this before starting:** that module does *not* use `eqdsk_file`. It calls `read_eqdsk` from `python/libneo/eqdsk_base.py`, which returns a plain dict and carries **no** COCOS sign guard, and it builds its spline directly in CGS (`R*METER_TO_CM`, `PsiVs.T*1.0e8`). So this step means constructing an `eqdsk_file(gfile)` alongside (or instead of) the existing `read_eqdsk` call and converting at the call site: `B_cgs = B_si * 1e4`, with R passed in metres and the resulting `dl`/`dR`/`dZ` still in cm. Since `b_at_rz` differentiates the same raw `PsiVs` and the sign guard only touches `_psi_at_edge` (not `PsiVs` itself), the poloidal field should agree — but verify rather than assume.
   Do **not** expect bit-identical output: rebuilding the spline in SI and rescaling changes floating-point rounding. Require `B_theta_cgs` to agree with the pre-change values to `1e-10` relative. A larger deviation is a real bug and belongs in this PR.
   If this turns out to be more than a mechanical swap, split it into a follow-up issue and land steps 1-6 and 8-10 alone; it is not load-bearing for the feature.

8. **`test/python/test_eqdsk_bfield.py`** (new) — implement the first four scenarios above against `test/resources/input_efit_file.dat`, following the class/fixture style of `test/python/test_eqdsk_interpolation.py`. No CMake registration is needed: `.github/workflows/main.yml:100` runs `python -m pytest test/python -q`.

9. **`python/tests/test_eqdsk_field_vs_fortran.py`** (new) — the fifth scenario, cross-checking against the Fortran. It must live in `python/tests/` rather than `test/python/` because only `python/tests/` has the `conftest.py` `libneo.__path__` shim (both directories carry `field_divB0.inp` and `efit_to_boozer.inp`, so the fixtures are not the deciding factor). Mark it `@pytest.mark.skipif` on `_efit_to_boozer` import failure so it degrades gracefully when the extension is not built. This test *is* the deliverable for libneo#33; cross-reference it there. `.github/workflows/main.yml:101` runs `python -m pytest python/tests -vv -s`.

10. **`docs/eqdsk.rst`** — add a "Magnetic Field" section with the three formulas and a worked example, and extend the `autoclass` `:members:` list (currently `psi_at_rz, spol_at_rz, theta_geometric_at_rz, rz_to_flux_coords`) with `dpsi_at_rz, fpol_at_rz, b_at_rz, bmod_at_rz, rz_at_flux_coords, b_at_flux_coords`. The Sphinx build is CI-gated (`.github/workflows/main.yml:131-151`, `sphinx-build -b html . _build/html`), so a typo in the member list fails the build.

## Acceptance criteria

- [ ] `python -m pytest test/python -q` passes, including the new `test/python/test_eqdsk_bfield.py`, with `test_eqdsk_interpolation.py` **unmodified**.
- [ ] `python -m pytest python/tests -vv -s` passes, including `test_eqdsk_field_vs_fortran.py` (or skipping it cleanly when the f2py extension is absent).
- [ ] On-axis: `abs(Bphi - Btor_at_R0*R0/Rpsi0) / abs(Btor_at_R0*R0/Rpsi0) < 2e-2`, and `hypot(BR, BZ) < 1e-2 * abs(Bphi)`.
- [ ] Outside the LCFS at `Z = Zpsi0`, `R in linspace(2.18, eq.R[-2], 20)`: `Bphi * R` equals `eq.fprof[-1]` to `1e-9` relative.
- [ ] `rz_at_flux_coords` → `rz_to_flux_coords` round-trip over `s_pol ∈ {0.1, 0.3, 0.6, 0.9}` × 17 angles: all 68 calls succeed, `|Δs_pol| < 1e-6`, wrapped `|Δtheta| < 1e-10`.
- [ ] `|B|` from Python within `1e-3` relative of `B0mod * 1e-4` from `python/efit_to_boozer/boozer.py:get_angles_G_and_B0` on the `s_tor = 0.5` surface.
- [ ] `python/tests/test_eqdsk_to_boozer_chartmap.py` still passes after the step-7 deduplication, and `B_theta_cgs` matches its pre-change values to `1e-10` relative.
- [ ] Fortran build and test unchanged (no Fortran source touched): `cmake -S . -B build -G Ninja -DCMAKE_BUILD_TYPE=Debug && cmake --build build --parallel && (cd build && ctest --output-on-failure)`.
- [ ] `cd docs && sphinx-build -b html . _build/html` succeeds with the extended `:members:` list.

## References

- L. L. Lao, H. St. John, R. D. Stambaugh, A. G. Kellman, W. Pfeiffer, *Reconstruction of current profile parameters and plasma shapes in tokamaks*, Nucl. Fusion **25** (1985) 1611 — the EFIT/G-EQDSK equilibrium whose `psi(R,Z)` and `F(psi)` are being differentiated here. The decomposition being implemented is the standard axisymmetric one, `B = F(psi) grad(phi) + grad(psi) × grad(phi)`.
- O. Sauter, S. Yu. Medvedev, *Tokamak coordinate conventions: COCOS*, Comput. Phys. Commun. **184** (2013) 293-302, doi:10.1016/j.cpc.2012.09.010 — the convention framework libneo already implements in `src/magfie/geqdsk_tools.f90`; read it before deciding what sign `b_at_rz` returns.
- W. D. D'haeseleer, W. N. G. Hitchon, J. D. Callen, J. L. Shohet, *Flux Coordinates and Magnetic Field Structure*, Springer (1991) — for the `s_pol` / `s_tor` / geometric-vs-straight-field-line angle distinctions that step 6 depends on.

Beyond these, the formulas in steps 2-4 are ordinary Grad-Shafranov algebra with no specific literature reference needed; they are an implementation detail internal to the code, and `src/magfie/field_divB0.f90:244-265` is the authoritative in-repo statement of them.

## Scope and risk

Single PR, roughly 250 lines of Python plus two test files.

The one real regression risk is step 1: swapping `RegularGridInterpolator(method='cubic')` for `RectBivariateSpline` changes interpolated `psi` at the 1e-6 level and changes *extrapolation* behaviour outside the grid — `bounds_error=False` currently yields `NaN`, whereas `RectBivariateSpline` extrapolates silently. Preserve the `NaN`-outside-box behaviour explicitly with a bounding-box mask, or `test_eqdsk_interpolation.py` and every downstream `spol_at_rz` caller quietly change meaning near the wall.

A second, smaller risk in step 5: `s_pol` is only guaranteed monotonic along a ray inside the LCFS. The `0.98 * l_max` bracket keeps `brentq` inside that region and was verified to work on the test fixture, but a diverted equilibrium with an X-point close to the axis could still break it. libneo#362 / itpplasma/efit_to_boozer#13 ("Get separatrix value from EQDSK file") would give a more robust bracket than the `Lcfs` polygon.

Related issues: libneo#33 (Fortran↔Python EQDSK consistency — step 9 is its deliverable), libneo#302 (g-file convention checks only on the `geqdsk_tools` path; `field_divB0` readers unvalidated), libneo#363 (automatically check for COCOS 3 in EQDSK input), libneo#362 / itpplasma/efit_to_boozer#13.
